### PR TITLE
Add support for Idris language

### PR DESCRIPTION
**Pull request by bmwant**
### Description

* Add lexer for [Idris](https://www.idris-lang.org/) programming language
* Ported from [sublime plugin tmLanguage](https://github.com/idris-hackers/idris-sublime/blob/master/Idris.YAML-tmLanguage)

### Testing

* Added specs
* Tested with visual/demo examples

image

### CODE REVIEW

**Comment by bmwant** [comment]
hey @pyrmont, could you take a look?

**Comment by pyrmont** [comment]
@bmwant Sorry not to have got around to this yet. I'm afraid I haven't had much time to spend on Rouge of late but will try to remedy that soon!

File: lib/rouge/lexers/idris.rb [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      aliases 'idr'
+      filenames '*.idr'
+      mimetypes 'text/x-idris'
+
+      reserved = %w(
+        _
+        data class instance namespace
+        infix[lr]? let where of with type 
+        do if then else case in
+      )
```
Comment (pyrmont) [inline lib/rouge/lexers/idris.rb]: Local variables defined in the scope of the class will not be garbage collected by Ruby and so have negative performance characteristics. For this reason, we try to encourage new lexers to use memoising methods instead (there are lexers in the repo that still use local variables but moving away from this is an ongoing proejct).

You can see an example in the C lexer for how to do this:

https://github.com/rouge-ruby/rouge/blob/81b36718a602d37b296301e5e9f4c8b2156b6ba2/lib/rouge/lexers/c.rb#L18-L27

The memoised value is then later referenced like this:

https://github.com/rouge-ruby/rouge/blob/81b36718a602d37b296301e5e9f4c8b2156b6ba2/lib/rouge/lexers/c.rb#L112-L127

File: lib/rouge/lexers/idris.rb [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+      ascii = %w(
+        NUL SOH [SE]TX EOT ENQ ACK BEL BS HT LF VT FF CR S[OI] DLE
+        DC[1-4] NAK SYN ETB CAN EM SUB ESC [FGRU]S SP DEL [abfnrtv\\\"'\&]
+      )
```
Comment (pyrmont) [inline lib/rouge/lexers/idris.rb]: See earlier comment.

File: lib/rouge/lexers/idris.rb [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        encodeFloat enumFrom enumFromThen enumFromThenTo enumFromTo exp
+        fail filter flip floor foldl foldl1 foldr foldr1 fromInteger fst
+        gcd getChar getLine head id init iterate last lcm length lines log
+        lookup map max maxBound maximum maybe min minBound minimum mod
+        negate not null or pi pred print product putChar putStr putStrLn
+        readFile recip repeat replicate return reverse scanl scanl1 sequence
+        sequence_ show sin sinh snd span splitAt sqrt succ sum tail take
+        takeWhile tan tanh uncurry unlines unwords unzip unzip3 words
+        writeFile zip zip3 zipWith zipWith3
+      )
```
Comment (pyrmont) [inline lib/rouge/lexers/idris.rb]: See earlier comment.

File: lib/rouge/lexers/idris.rb [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      state :root do
+        mixin :basic
+        mixin :directive
+
+        rule %r/\bimport\b/, Keyword::Reserved, :import
+        rule %r/\bmodule\b/, Keyword::Reserved, :module
+        rule %r/\b(?:#{reserved.join('|')})\b/, Keyword::Reserved
+        rule %r/\b(Just|Nothing|Left|Right|True|False|LT|LTE|EQ|GT|GTE)\b/, Keyword::Constant
+        # function signature
+        rule %r/^([\w']+)\s*(:)/, Name::Function
```
Comment (pyrmont) [inline lib/rouge/lexers/idris.rb]: Is there any reason to use capture groups for `[\w']` and `:`? Is it just for legibility?

File: lib/rouge/lexers/idris.rb [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        rule %r(\\(?![:!#\$\%&*+.\\/<=>?@^\|~-]+)), Name::Function
+        # special operators
+        rule %r((<-|::|->|=>|=)(?![:!#\$\%&*+.\\/<=>?@^\|~-]+)), Operator
+        # constructor/type operators
+        rule %r(:[:!#\$\%&*+.\\/<=>?@^\|~-]*), Operator
+        # other operators
+        rule %r([:!#\$\%&*+.\\/<=>?@^\|~-]+), Operator
+
+        rule %r/\d+e[+-]?\d+/i, Num::Float
+        rule %r/\d+\.\d+(e[+-]?\d+)?/i, Num::Float
```
Comment (pyrmont) [inline lib/rouge/lexers/idris.rb]: Could you combine these rules like this?

```suggestion
        rule %r/\d+(\.\d+)?(e[+-]?\d+)?/i, Num::Float
```

**Comment by tancnle** [comment]
👋🏼 @bmwant we would love to hear back from you. Is there anything we can help to push this PR forward?

**Comment by bmwant** [comment]
@tancnle hey, I can recall that I've faced a couple of blocking issues when trying to address these simple at a first glance comments. I cannot remember whether it was related to the Ruby language (as I'm not a Ruby developer per se) or to the implementation details of the lexer itself (regex lexers are trickier to implement).
If that's something related to the language I'll reach back here with my questions/clarifications, but in case it's the latter (and I suspect it's more likely) I just don't have enough mental energy to face again that horrible process of debugging 😢

**Comment by tancnle** [comment]
@bmwant Thank you for getting back to me.  I am happy to lend a hand with the latter. Idris is an interesting language and this addition to Rouge will definitely improve the exp. for our developers. 

I will review those comments from `@pyrmont` and help to address them in the meantime.

**Comment by nickdrozd** [comment]
Would it be possible to add a simpler version of Idris highlighting first, and then polish it later? I've been using Haskell highlighting for my Idris code blocks, and that looks terrible, so for me any real Idris support at all would be an improvement.

(I'm afraid I don't have anything concrete to offer in terms of this PR. This is just a suggestion to take the simpler-sooner route.)

File: lib/rouge/lexers/idris.rb (hunk 4 above)
Comment (tancnle) [reply lib/rouge/lexers/idris.rb]: Addressed in https://github.com/rouge-ruby/rouge/pull/1464/commits/3a31bf6eb5dd8193f34b2de9cefea6d06d2768e3

File: lib/rouge/lexers/idris.rb (hunk 5 above)
Comment (tancnle) [reply lib/rouge/lexers/idris.rb]: Addressed in https://github.com/rouge-ruby/rouge/pull/1464/commits/5e638d46c828f3eb3ed01202ca5f52a79c583d34

File: lib/rouge/lexers/idris.rb (hunk 1 above)
Comment (tancnle) [reply lib/rouge/lexers/idris.rb]: Addressed in https://github.com/rouge-ruby/rouge/pull/1464/commits/d62d75e39180b4c87f5f64fa122918acfa6a6fe5

File: lib/rouge/lexers/idris.rb (hunk 3 above)
Comment (tancnle) [reply lib/rouge/lexers/idris.rb]: Addressed in https://github.com/rouge-ruby/rouge/pull/1464/commits/d62d75e39180b4c87f5f64fa122918acfa6a6fe5

File: lib/rouge/lexers/idris.rb (hunk 2 above)
Comment (tancnle) [reply lib/rouge/lexers/idris.rb]: Addressed in https://github.com/rouge-ruby/rouge/pull/1464/commits/d62d75e39180b4c87f5f64fa122918acfa6a6fe5

**Comment by tancnle** [comment]
@bmwant I have pushed some changes to address the feedback from @pyrmont. Could you please have another look? 🙏🏼

**Comment by tancnle** [comment]
@bmwant In the spirit of iteration, I will merge this change 👍🏼. Please let me know if you have any concerns and/or raise another PR to address them.

**Pull request merged**